MIP/mip-61/README.md

```diff
@@ -0,0 +1,185 @@
+
+!!! . Only finalized source blocks should be considered. So there is need for implementing a trigger for when a block is finalized.
+
+![alt text](continuous_processing.png)
+
+1. Receive new non-processed finalized `source_block`.
+1. If a block height is missing put the block into a queue. DO NOT process blocks out of order.
+1. Get `initiate_transfer` events from the `source_block`.
+1. For each `initiate_transfer` event run the event_processing script.
+1. Set the block height as `block_processed`.
+
+**Optimizations to consider errors in the above procedure**
+This step is optional but should be considered.
+
+Introduce a status `transfer_init` to differentiate a state between nonce creation locally and sending successfully a `complete_transfer` transaction.
+
+![alt text](init_optimization.png)
+
+**Calculation-Completed-Block-Height**
+In this section the process is defined to calculate the completed part of the source chain `completed_block_height` and `completed_nonce_height`.
+
+![alt text](complete_block_height.png)
+
+1. If completed set status transfer as `transfer_completed` and return.
+
+is replaced by
+
+1. If completed set status transfer as `transfer_completed`.
+1. If the nonce is not `completed_nonce_height + 1` then return.
+1. Else
+    1. Set `completed_nonce_height += 1`.
+    1. If this nonce is the last nonce in the source block set `completed_block_height += 1`.
+    1. If the next nonce is completed start from step 1.
+    1. Else return.
+
+### Timeout algorithm
+
+1. Whenever the timeout of a transfer (which has `transfer_pending` status) is triggered. Start the event processing.
+
+
+![alt text](timeout.png)
+
+### Bootstrapping
+
+Next we describe how the bootstrap algorithm works and differs from the above.
+
+!!! . A node that is bootstrapping MUST start the [Continuous-Processing](#continuous-processing) algorithm in parallel.
+
+![alt text](bootstrap.png)
+
+The Algorithm differs from the Continuous-Processing in that it runs in parallel and will catch up with missing transfers eventually. While not hindering the Continuous Operation of the Relayer. It implements a delay to start at the beginning, which conveniently prevents that `complete_transfer` transactions would be sent accidentally twice to the target chain.
+
+1. Start the Continuous-Processing protocol.
+1. Set the first processed source block by the Continuous-Processing protocol as `first_continuous_processed_block`.
+1. Set `end_source_block = first_continuous_processed_block - 1`
+1. Set `start_source_block` according to some input.
+1. Wait for some time `wait_time`, e.g. 10 minutes. This is to ensure that all relevant source blocks have arrived. It may be not necessary but it also does not hurt.
+1. For each source block between `start_source_block` and `end_source_block` perform the same algorithm as Continuous-Processing (apart from using the next source block height in the loop rather than a new source block)
```

Comment:
If we have a non-decreasing nonce, can't we binary search this?

Reply:
nonces could be skipped temporarily on the target chain, thus we have to go through it one-by-one